### ISSUE TITLE
string comparision causes silent malfunction

### ISSUE THREAD

**Issue by navytux** [iovisor/bpftrace#402]
Hello up there. I've discovered that if I use `==` for strings in a probe, even if it should take no effect, the whole probe starts to malfunction. For example consider the following program:

```c
#include <linux/path.h>
#include <linux/dcache.h>
#include <linux/fs.h>

kprobe:vfs_write {
    $file = (file *)arg0;
    $size = arg2;
    $dentry = $file->f_path.dentry;
    $file_name = str($dentry->d_name.name);

    //if ($file_name == "zzz") {
    //    $size = $size;
    //}

    @writes[$file_name] = $size;
}
```

if I do `cat </dev/zero >/tmp/aaa` while `bpftrace test.bt` runs, I get something like:

```
root@deco:# bpftrace test.bt
Attaching 1 probe...
^C

@writes[]: 1
@writes[.org.chromium.Chromium.O55K1k]: 515328
@writes[2]: 30
@writes[[eventfd]]: 8
@writes[aaa]: 98304                 <-- NOTE
@writes[ptmx]: 1
@writes[tun]: 80
```

All as expected. However if I uncomment the block which compares `$file_name` to `"zzz"`

```c
    if ($file_name == "zzz") {
        $size = $size;
    }
```
which should not change the semantic (even if it is "zzz", it is only idempotent `$size = $size` assignment), observed behaviour under the same conditions become:

```
root@deco:# bpftrace test.bt
Attaching 1 probe...
^C

@writes[]: 1
@writes[ntfdntfd]]: 8
@writes[og]: 79
@writes[on.lon.log]: 79
@writes[tnettnet.log]: 108
```

i.e. no "aaa" at all and other names seem to be corrupted.

The difference in between generated programs (`bpftrace -d`) is attached below.

Thanks beforehand,
Kirill

---- 8< ----

```diff
--- a/da.txt
+++ b/db.txt
@@ -28,6 +28,14 @@ Program
        variable: $dentry
       d_name
      name
+  if
+   ==
+    variable: $file_name
+    string: zzz
+   then
+    =
+     variable: $size
+     variable: $size
   =
    map: @writes
     variable: $file_name
@@ -48,49 +56,89 @@ define i64 @"kprobe:vfs_write"(i8*) local_unnamed_addr section "s_kprobe:vfs_wri
 entry:
   %"@writes_val" = alloca i64, align 8
   %"@writes_key" = alloca [64 x i8], align 1
+  %strcmp.char14 = alloca i8, align 1
+  %strcmp.char10 = alloca i8, align 1
+  %strcmp.char6 = alloca i8, align 1
+  %strcmp.char = alloca i8, align 1
+  %"$file_name" = alloca [64 x i8], align 1
+  %1 = getelementptr inbounds [64 x i8], [64 x i8]* %"$file_name", i64 0, i64 0
+  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %1)
+  call void @llvm.memset.p0i8.i64(i8* nonnull align 1 %1, i64 0, i64 64, i1 false)
   %qstr.name = alloca i64, align 8
   %str = alloca [64 x i8], align 1
   %path.dentry = alloca i64, align 8
   %arg2 = alloca i64, align 8
   %arg0 = alloca i64, align 8
-  %1 = bitcast i64* %arg0 to i8*
+  %2 = bitcast i64* %arg0 to i8*
+  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %2)
+  %3 = getelementptr i8, i8* %0, i64 112
+  %probe_read = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i64* nonnull %arg0, i64 8, i8* %3)
+  %4 = load i64, i64* %arg0, align 8
+  call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %2)
+  %5 = bitcast i64* %arg2 to i8*
+  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %5)
+  %6 = getelementptr i8, i8* %0, i64 96
+  %probe_read1 = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i64* nonnull %arg2, i64 8, i8* %6)
+  %7 = load i64, i64* %arg2, align 8
+  call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %5)
+  %8 = add i64 %4, 24
+  %9 = bitcast i64* %path.dentry to i8*
+  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %9)
+  %probe_read2 = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i64* nonnull %path.dentry, i64 8, i64 %8)
+  %10 = load i64, i64* %path.dentry, align 8
+  call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %9)
+  %11 = getelementptr inbounds [64 x i8], [64 x i8]* %str, i64 0, i64 0
+  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %11)
+  call void @llvm.memset.p0i8.i64(i8* nonnull align 1 %11, i8 0, i64 64, i1 false)
+  %12 = add i64 %10, 40
+  %13 = bitcast i64* %qstr.name to i8*
+  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %13)
+  %probe_read3 = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i64* nonnull %qstr.name, i64 8, i64 %12)
+  %14 = load i64, i64* %qstr.name, align 8
+  call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %13)
+  %probe_read_str = call i64 inttoptr (i64 45 to i64 (i8*, i64, i8*)*)([64 x i8]* nonnull %str, i64 64, i64 %14)
   call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %1)
-  %2 = getelementptr i8, i8* %0, i64 112
-  %probe_read = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i64* nonnull %arg0, i64 8, i8* %2)
-  %3 = load i64, i64* %arg0, align 8
+  call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull align 1 %1, i8* nonnull align 1 %11, i64 64, i1 false)
+  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %strcmp.char)
+  %probe_read4 = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i8* nonnull %strcmp.char, i64 8, [64 x i8]* nonnull %"$file_name")
+  %15 = load i8, i8* %strcmp.char, align 1
+  %strcmp.cmp = icmp eq i8 %15, 122
+  br i1 %strcmp.cmp, label %strcmp.loop, label %strcmp.false
+
+strcmp.false:                                     ; preds = %strcmp.loop9, %strcmp.loop5, %strcmp.loop, %entry
   call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %1)
-  %4 = bitcast i64* %arg2 to i8*
-  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %4)
-  %5 = getelementptr i8, i8* %0, i64 96
-  %probe_read1 = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i64* nonnull %arg2, i64 8, i8* %5)
-  %6 = load i64, i64* %arg2, align 8
-  call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %4)
-  %7 = add i64 %3, 24
-  %8 = bitcast i64* %path.dentry to i8*
-  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %8)
-  %probe_read2 = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i64* nonnull %path.dentry, i64 8, i64 %7)
-  %9 = load i64, i64* %path.dentry, align 8
-  call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %8)
-  %10 = getelementptr inbounds [64 x i8], [64 x i8]* %str, i64 0, i64 0
-  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %10)
-  call void @llvm.memset.p0i8.i64(i8* nonnull align 1 %10, i8 0, i64 64, i1 false)
-  %11 = add i64 %9, 40
-  %12 = bitcast i64* %qstr.name to i8*
-  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %12)
-  %probe_read3 = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i64* nonnull %qstr.name, i64 8, i64 %11)
-  %13 = load i64, i64* %qstr.name, align 8
-  call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %12)
-  %probe_read_str = call i64 inttoptr (i64 45 to i64 (i8*, i64, i8*)*)([64 x i8]* nonnull %str, i64 64, i64 %13)
-  %14 = getelementptr inbounds [64 x i8], [64 x i8]* %"@writes_key", i64 0, i64 0
-  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %14)
-  call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull align 1 %14, i8* nonnull align 1 %10, i64 64, i1 false)
-  %15 = bitcast i64* %"@writes_val" to i8*
-  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %15)
-  store i64 %6, i64* %"@writes_val", align 8
+  %16 = getelementptr inbounds [64 x i8], [64 x i8]* %"@writes_key", i64 0, i64 0
+  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %16)
+  call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull align 1 %16, i8* nonnull align 1 %1, i64 64, i1 false)
+  %17 = bitcast i64* %"@writes_val" to i8*
+  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %17)
+  store i64 %7, i64* %"@writes_val", align 8
   %pseudo = call i64 @llvm.bpf.pseudo(i64 1, i64 1)
   %update_elem = call i64 inttoptr (i64 2 to i64 (i8*, i8*, i8*, i64)*)(i64 %pseudo, [64 x i8]* nonnull %"@writes_key", i64* nonnull %"@writes_val", i64 0)
-  call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %14)
+  call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %16)
   ret i64 0
+
+strcmp.loop:                                      ; preds = %entry
+  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %strcmp.char6)
+  %18 = add [64 x i8]* %"$file_name", i64 1
+  %probe_read7 = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i8* nonnull %strcmp.char6, i64 8, [64 x i8]* %18)
+  %19 = load i8, i8* %strcmp.char6, align 1
+  %strcmp.cmp8 = icmp eq i8 %19, 122
+  br i1 %strcmp.cmp8, label %strcmp.loop5, label %strcmp.false
+
+strcmp.loop5:                                     ; preds = %strcmp.loop
+  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %strcmp.char10)
+  %20 = add [64 x i8]* %"$file_name", i64 2
+  %probe_read11 = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i8* nonnull %strcmp.char10, i64 8, [64 x i8]* %20)
+  %21 = load i8, i8* %strcmp.char10, align 1
+  %strcmp.cmp12 = icmp eq i8 %21, 122
+  br i1 %strcmp.cmp12, label %strcmp.loop9, label %strcmp.false
+
+strcmp.loop9:                                     ; preds = %strcmp.loop5
+  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %strcmp.char14)
+  %22 = add [64 x i8]* %"$file_name", i64 3
+  %probe_read15 = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i8* nonnull %strcmp.char14, i64 8, [64 x i8]* %22)
+  br label %strcmp.false
 }
 
 ; Function Attrs: argmemonly nounwind
```

**Comment by brendangregg** [comment]
This could be #198, which is a priority to fix.

**Comment by navytux** [comment]
@brendangregg, thanks for feedback. I confirm that #405 (the fix for #198) fixes this issue as well (cc @williangaspar).

**Comment by navytux** [comment]
> I confirm that #405 (the fix for #198) fixes this issue as well

Though not fully - in the following program the check for `$file_name` and map update `$@writes[$file_name] = $size` are working correctly, but printf under check misbehaves:

```c
#include <linux/path.h>
#include <linux/dcache.h>
#include <linux/fs.h>

kprobe:vfs_write {
    $file = (file *)arg0;
    $size = arg2;
    $dentry = $file->f_path.dentry;
    $file_name = str($dentry->d_name.name);

    if ($file_name == "aaa") {
        printf("aaa (%s)\n", $file_name);
    }

    @writes[$file_name] = $size;
}
```

it prints:

```
aaa ()
aaa ()
aaa ()
aaa ()
...
```

but it must be

```
aaa (aaa)
aaa (aaa)
aaa (aaa)
aaa (aaa)
...
```

**Comment by navytux** [comment]
Thanks.
